oak/server/rust/oak_runtime/src/metrics/mod.rs

```diff
@@ -15,53 +15,142 @@
+}
+
+/// Mostly copied from https://github.com/grpc-ecosystem/java-grpc-prometheus
+#[derive(Clone)]
+pub struct GrpcServerMetrics {
+    /// Total number of RPCs started on the server.
+    pub grpc_server_started_total: IntCounterVec,
+    /// Total number of RPCs completed on the server, regardless of success or failure.
+    pub grpc_server_handled_total: IntCounterVec,
+    /// Histogram of response latency of RPCs handled by the server, in seconds.
+    pub grpc_server_handled_latency_seconds: HistogramVec,
+    /// Histogram of response sizes of RPCs handled by the server.
+    pub grpc_response_size: HistogramVec,
+
+    /// Total number of stream messages received from the client.
+    pub grpc_server_msg_received_total: IntCounterVec,
+    /// Total number of stream messages sent by the server.
+    pub grpc_server_msg_sent_total: IntCounterVec,
+}
+
+/// Struct that collects all metrics for monitoring the Oak Runtime.
+#[derive(Clone)]
+pub struct RuntimeMetrics {
+    pub runtime_nodes_count: IntGaugeVec,
+}
+
 /// Struct that collects all the metrics in one place
+#[derive(Clone)]
 pub struct Metrics {
-    pub grpc_request_duration: Histogram,
-    pub grpc_requests_total: IntCounter,
-    pub grpc_response_size: Histogram,
-    pub runtime_nodes_count: IntGauge,
+    registry: Registry,
+    pub runtime_metrics: RuntimeMetrics,
+    pub grpc_server_metrics: GrpcServerMetrics,
 }
 
-// TODO(#899): For testability implement a trait with methods for updating the metrics.
-// TODO(#899): Instead of using a global Registry, the Runtime should instantiate and manage the
-// Registry
-impl Metrics {
-    fn new() -> Self {
+impl MetricsBuilder {
+    pub fn new() -> Self {
         Self {
-            grpc_request_duration: register_histogram!(
-                "grpc_request_duration_seconds",
-                "The gRPC request latencies in seconds."
-            )
-            .expect("Creating grpc_request_duration_seconds metric failed."),
-
-            grpc_requests_total: register_int_counter!(opts!(
-                "grpc_requests_total",
-                "Total number of gRPC requests received."
-            ))
-            .expect("Creating grpc_requests_total metric failed."),
-
-            grpc_response_size: register_histogram!(
-                "grpc_response_size_bytes",
-                "The gRPC response sizes in bytes."
-            )
-            .expect("Creating grpc_response_size_bytes metric failed."),
-
-            runtime_nodes_count: register_int_gauge!(opts!(
+            registry: Registry::new(),
+        }
+    }
+
+    fn register<T: 'static + prometheus::core::Collector + Clone>(&self, metric: T) -> T {
+        self.registry.register(Box::new(metric.clone())).unwrap();
+        metric
+    }
+
+    fn counter(&self, metric_name: &str, labels: &[&str], help: &str) -> IntCounterVec {
+        let opts = Opts::new(metric_name, help);
+        IntCounterVec::new(opts, labels).unwrap()
+    }
```

Comment:
AFAICT these methods don't do anything with `self`, in fact I am not sure whether they need to be defined in this `impl` block in the first place, is this an intentional decision?